File: library/src/main/java/com/carlosmuvi/segmentedprogressbar/DrawingTimer.java
package com.carlosmuvi.segmentedprogressbar;

import android.os.Handler;

public class DrawingTimer {

    private static final long TICK_TIME_MILLISECONDS = 10;

    private final Handler handler;
    private int totalTicks;
    private int currentTick = 0;
    private Listener listener;
    private TimerState timerState = TimerState.IDLE;

    public DrawingTimer() {
        handler = new Handler();
    }

    public void start(long timeInMilliseconds) {
        requireMinimumTickTime(timeInMilliseconds);
        if (timerState == TimerState.IDLE) {
            this.totalTicks = (int) (timeInMilliseconds / TICK_TIME_MILLISECONDS);
        }
        if (timerState != TimerState.RUNNING) {
            timerState = TimerState.RUNNING;
            runDrawingTask();
        }
    }

    private void requireMinimumTickTime(long timeInMilliseconds) {
        if (timeInMilliseconds < TICK_TIME_MILLISECONDS) {
            String errorMessage = "A minimum of " +
                    TICK_TIME_MILLISECONDS +
                    " milliseconds is required, but input is " +
                    timeInMilliseconds + " milliseconds.";
            throw new IllegalArgumentException(errorMessage);
        }
    }

    private void runDrawingTask() {
        handler.post(new Runnable() {
            @Override
            public void run() {
                listener.onTick(currentTick, totalTicks);
                currentTick++;
                if (currentTick <= totalTicks) {
                    handler.postDelayed(this, TICK_TIME_MILLISECONDS);
                } else {
                    reset();
                }
            }
        });
    }

    public void pause() {
        if (timerState == TimerState.RUNNING) {
            timerState = TimerState.PAUSED;
            handler.removeCallbacksAndMessages(null);
        }
    }

    public void resume() {
        if (timerState == TimerState.PAUSED) {
            timerState = TimerState.RUNNING;
            runDrawingTask();
        }
    }

    public void reset() {
        pause();
        timerState = TimerState.IDLE;
        currentTick = 0;
    }

    public void setListener(Listener listener) {
        this.listener = listener;
    }

    public boolean isRunning() {
        return timerState == TimerState.RUNNING;
    }

    public boolean isPaused() {
        return timerState == TimerState.PAUSED;
    }

    enum TimerState {
        RUNNING, PAUSED, IDLE
    }

    interface Listener {
        void onTick(int currentTicks, int totalTicks);
    }
}

File: library/src/main/java/com/carlosmuvi/segmentedprogressbar/DrawingTimer.kt
package com.carlosmuvi.segmentedprogressbar

import android.os.Handler

/**
 * Created by carlosmuvi on 02/09/16.
 */

class DrawingTimer(var listener: ((Int, Int) -> Unit)?) {

  private val handler: Handler = Handler()
  private val tickTimeInMilliseconds: Long = 30
  private var totalTicks: Int = 0
  private var currentTick = 0
  private var timerState = TimerState.IDLE

  fun start(timeInMilliseconds: Long) {
    if (timerState == TimerState.IDLE) {
      this.totalTicks = (timeInMilliseconds / tickTimeInMilliseconds).toInt()
    }
    if (timerState != TimerState.RUNNING) {
      timerState = TimerState.RUNNING
      runDrawingTask()
    }
  }

  private fun runDrawingTask() {
    handler.post(object : Runnable {
      override fun run() {
        listener?.invoke(currentTick, totalTicks)
        currentTick++
        if (currentTick <= totalTicks) {
          handler.postDelayed(this, tickTimeInMilliseconds)
        } else {
          reset()
        }
      }
    })
  }

  fun pause() {
    if (timerState == TimerState.RUNNING) {
      timerState = TimerState.PAUSED
      handler.removeCallbacksAndMessages(null)
    }
  }

  fun resume() {
    if (timerState == TimerState.PAUSED) {
      timerState = TimerState.RUNNING
      runDrawingTask()
    }
  }

  fun reset() {
    pause()
    timerState = TimerState.IDLE
    currentTick = 0
  }

  val isRunning: Boolean
    get() = timerState == TimerState.RUNNING

  val isPaused: Boolean
    get() = timerState == TimerState.PAUSED

  internal enum class TimerState {
    RUNNING, PAUSED, IDLE
  }
}
